lighthouse-core/gather/gatherers/link-elements.js

```diff
@@ -6,31 +6,111 @@
 'use strict';
 
 const Gatherer = require('./gatherer.js');
-const getElementsInDocumentString = require('../../lib/page-functions.js').getElementsInDocumentString; // eslint-disable-line max-len
+const URL = require('../../lib/url-shim.js').URL;
+const NetworkAnalyzer = require('../../lib/dependency-graph/simulator/network-analyzer.js');
+const LinkHeader = require('http-link-header');
+const getElementsInDocumentString = require('../../lib/page-functions.js')
+  .getElementsInDocumentString; // eslint-disable-line max-len
+
+/**
+ *
+ * @param {string} url
+ * @param {string} finalUrl
+ * @return {string|null}
+ */
+function normalizeUrlOrNull(url, finalUrl) {
+  try {
+    return new URL(url, finalUrl).href;
```

Comment:
accessing the `.href` of a link from dom already does the normalization for us

we do call attention to these cases when it matters in the audits themselves, i.e. there was an `hrefRaw` but no `href` indicates there was an error in parsing the URL

Reply:
> we do call attention to these cases when it matters in the audits themselves, i.e. there was an hrefRaw but no href indicates there was an error in parsing the URL

I was hoping we could do more to make it less tricky to get right for any audits using the artifact (e.g. boolean props for the different invalid permutations rather than making sure to check both href and hrefRaw), but I don't have a good suggestion, so let's wait for if it's ever an issue :)